---
title: "Ginty lab Neuron RNA-Seq Analysis: Follow-up"
output:
  html_document:
    theme: cerulean
    toc: true
    toc_depth: 4
    fig_caption: true
    fig_width: 8
    fig_height: 6
author: "Meeta Mistry"

---

```{r setup, echo=FALSE}
 
# Setup report details
clientname="Yang Zheng"
clientemail="zylittlep@gmail.com"
lablocation="Neurobiology"
analystname="Meeta Mistry"
analystemail="mmistry@hsph.harvard.edu"
```

RNA-Seq analysis for `r clientname` (`r clientemail`) at `r lablocation`. Contact `r analystname` (`r analystemail`) for additional details. Follow up with the client:     

* Evaluate expression patterns of the five marker genes used for FACS sorting to validate specifity
* Also look at the handful of 'neuron-specific' genes identified through Galaxy analysis
* Interested in genes only enriched in LTMR populations using Nociceptor populations as background and vice versa. Can we extract gene lists for each


## Setup

### Bioconductor and R libraries used

```{r libraries, echo=TRUE}
loadlibs <- function(){
library(ggplot2)
library(gplots)
library(CHBUtils)
library(grid)
library(gridExtra)
library(Biobase)
library(genefilter)
library(RColorBrewer)
library(edgeR)
library(DESeq2)
}
suppressPackageStartupMessages(loadlibs())
```

### Get variables
- get base directory for analyses
- specify data and results directories
- specify column headers used in metadata file

```{r directories, echo=TRUE}
baseDir=getwd()
dataDir=paste(baseDir, "/data", sep="")
resultsDir=paste(baseDir, "/results", sep="")
metaDir=paste(dataDir, "/meta", sep="")

heatcolors.1 <- rev(brewer.pal(6, "YlOrRd"))
heatcolors.2 <- colorRampPalette(brewer.pal(9, "GnBu"))(100)
cbPalette <- cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00")
```

## Load data and annotations
Source for gtf annotation file found at [GENCODE](http://www.gencodegenes.org/mouse_releases/). Combine data + metadata + feature data into an _eset_ object.

```{r load data, echo=TRUE}

# Load data and metadata
counts <- read.delim(file.path(dataDir, 'annotated_combined.counts'), sep="\t", row.names=1, header=T, as.is=T)
meta <- read.delim(file.path(metaDir, 'ginty_neuron.tsv'), sep="\t", row.names=1, header=T)

# extract gene symbols
symbol <- counts$symbol
counts <- counts[,which(colnames(counts) != "symbol")]

# extract ensembl IDs
ensembl <- rownames(counts)

# Remove outlier sample
outlier <- "Npy2r3"
meta <- meta[which(row.names(meta) != outlier),]
counts <- counts[,which(colnames(counts) %in% rownames(meta))]

# match annotations with count data
annot <- data.frame(symbol, row.names=ensembl, stringsAsFactors=FALSE)

# Create eset object
eset <- new("ExpressionSet", exprs=as.matrix(counts))
meta <- meta[colnames(counts),]
pData(eset) <- meta
fData(eset) <- annot

```

## Look at marker genes for validation
The following marker genes were used  in FACS sorting and therefore should show a high degree of specifity. Calcb is the only gene that doesn't quite agree.        

* Mrgprd: Nonpeptidergic
* Calca, Calcb: Peptidergic
* Pvalb: Propriocpetor
* Th: C-LTMR

```{r markergenes, echo=FALSE, fig.align='center', fig.height=20, fig.width=20, warning=FALSE}

# Get exclusive genes
markers <- c("Mrgprd", "Calca", "Calcb", "Pvalb", "Th")
color <- brewer.pal(n=7, name="Set3")
exp <- cpm(exprs(eset))

# stripchart
# Set up a blank stripchart
par(las=2, mfrow=c(4,2), mar=c(2,4,2,2), cex=1.5)

for (i in markers){
    stripchart(
			exp[which(fData(eset)$symbol %in% i),] ~ eset$neurontype,
			method="jitter", vertical=TRUE,
			ylab="Expression (cpm)", col=NA,
			xlim=c(0.25,nlevels(eset$neurontype)+0.75),
			ylim=range(exp[which(fData(eset)$symbol %in% i),]),
			main= i,
      xaxt="n"
		);
		# Then draw points for each sample
		for (n in 1:nlevels(eset$neurontype)) {
			j <- which(eset$neurontype == levels(eset$neurontype)[n]);
			stripchart(
				exp[which(fData(eset)$symbol %in% i), j],
				method="jitter", vertical=TRUE, add=TRUE, at=n,
				pch=21, cex=1.5, bg=color[n]
			);
		}
}
plot.new()
legend("topleft", inset=0, title="Neurontype", legend=levels(pData(eset)$neurontype), fill=color, horiz=FALSE, cex=1.2)
```

## DESeq2 Gene Lists

* Neuron specific gene lists found using DESeq2 (neurontype vs all neuron background):
  * [Aβ.Nociceptor.WDR](./DESeq2_results/neurontypeAβ.Nociceptor..WDR)
  * [Aβ.RA.LTMR](./DESeq2_results/neurontypeAβ.RA.LTMR)
  * [Aδ.LTMR](./DESeq2_results/neurontypeAδ.LTMR)
  * [C.LTMR](./DESeq2_results/neurontypeC.LTMR)
  * [Nonpeptidergic.Nociceptor](./DESeq2_results/neurontypeNonpeptidergic.Nociceptor)
  * [Peptidergic.Nociceptor](./DESeq2_results/neurontypePeptidergic.Nociceptor)
  * [Proprioceptor](./DESeq2_results/neurontypeProprioceptor)

## Galaxy defined neuron-specific expression
The genes identified by hand from Galaxy data are few; a handful per neurontype. Of these genes the majority overlap with the gene lists generated by DESeq2 (as shown in the table below). The heatmap below illustrates the expression specificity (unit of expression here is once again using cpm).
```{r neurongenes, fig.align='center', results='asis'}

# Get Galaxy genes
neuron.genes <- read.delim(file.path(dataDir, 'gene lists/neuron.specific.txt'), header=T, sep="\t", row.names=1)

# Read in previous gene lists and compute overlaps
file.list <- read.delim("DESeq2_results/resultfiles.txt", header=F, as.is=T)
overlaps <- as.data.frame(matrix(0, ncol = 3, nrow = 7))
for (f in 1:nrow(file.list)){
  file <- file.list[f,1]
  x <- read.delim(paste("DESeq2_results/", file, sep=""), header=T, sep="\t", as.is=T)
  ids <- x[,1]
  genes <- fData(eset)[ids,]
  galaxy.genes <- row.names(neuron.genes)[grep(gsub("neurontype", "", file), neuron.genes$neurontype)]
  o <- length(which(galaxy.genes %in% genes))
  overlaps[f,] <- c(gsub("neurontype", "", file), length(galaxy.genes), o)
}

colnames(overlaps) <- c("Neurontype", "Galaxy genes", "Genes overlapping with DESeq2")
kable(overlaps, format='markdown')

# get expression data for genes
select <- match(row.names(neuron.genes), fData(eset)$symbol)

# set colors and genes
colors <- cbPalette[neuron.genes$neurontype]
colors <- colors[!is.na(select)]
select <- select[!is.na(select)]

# plot heatmap
heatmap.2(exprs(eset)[select,],scale="row", col=heatcolors.2, labRow=fData(eset)$symbol[select], Colv=FALSE, Rowv=FALSE, 
          RowSideColors=colors, ColSideColors=cbPalette[pData(eset)$neurontype], trace="none", dendrogram="none")
```


## Finding LTMR and Nociceptor specific genes
For each neurontype we can set the contrasts such that comparisons are made against a higher level 'group of neurons' rather than every other neurontype. Problem is we get alot of NA's for padj. 

_This was also attempted by using a general 'neuron group' factor, but the problem with this experimental design is that all of one neurontype is in one neuron group, so you can't separate the effect of these two (i.e. linearly dependent covariates)_

```{r deseq2-setup, eval=FALSE}

# Remove unwanted samples
sub.eset <- eset[,which(pData(eset)$neurontype != 'Proprioceptor')]
pData(sub.eset) <- droplevels(pData(sub.eset))

# Create DESeq2 dataset
se <-SummarizedExperiment(assays=SimpleList(counts=exprs(sub.eset)), colData=DataFrame(pData(sub.eset)))
dds <- DESeqDataSet(se=se, design = ~ neurontype)
dds <- DESeq(dds)

# Create list object for the results 
resultsAll <- vector("list", 6)

# adding the effects across all neurontypes within a group to create background contrast
bg.ltmr <- resultsNames(dds)[grep("LTMR", resultsNames(dds))]
bg.nociceptor <- resultsNames(dds)[grep("Nociceptor", resultsNames(dds))]

# Extract data of specified contrasts with appropriate background
for (n in 2:length(resultsNames(dds))){
  r <- resultsNames(dds)[n] 
  if (r %in% bg.ltmr) contrast <- list(r, c(bg.nociceptor))
  if (r %in% bg.nociceptor) contrast <- list(r, c(bg.ltmr))
  res <- results(dds, contrast=contrast)
  resultsAll[n-1] <- list(res)
}
names(resultsAll) <- resultsNames(dds)[-1]

```



