---
title: "Ginty lab Neuron RNA-Seq - Combined Data"
output:
  html_document:
    theme: cosmo
    toc: true
    toc_depth: 4
    fig_caption: true
    fig_width: 8
    fig_height: 6
author: "Meeta Mistry"
date: "04/29/2014"
---

```{r setup, echo=FALSE}
 
# Setup report details
clientname="Yang Zheng"
clientemail="zylittlep@gmail.com"
lablocation="Neurobiology"
analystname="Meeta Mistry"
analystemail="mmistry@hsph.harvard.edu"
```

RNA-Seq analysis for `r clientname` (`r clientemail`) at `r lablocation`. Contact `r analystname` (`r analystemail`) for additional details. Request from client was:     

> We consulted with the bioinformatics core for RNA-seq analysis a while ago. I just got my new RNA-seq data and I'm wondering whether we can re-open the consult and add these samples in for analysis. Additional samples include:
>
> * Abeta-RA-LTMR:NPY2R150407, NPY2R150516, NPY2R750820
> * Abeta-SA1-LTMR: SA1-2, SA1-04080419, SA1-04200521


## Workflow:   
  * run fastq files through [bcbio](https://bcbio-nextgen.readthedocs.org/en/latest/index.html)
  * [assess QC](https://dl.dropboxusercontent.com/u/35207958/ginty_neuron_2/combined-data/qc-summary-combined.html) 
  * Correlation structure / clustering (inter, intra)
  * Looking for condition-specific genes (one versus all comparisons)

## Setup

### Bioconductor and R libraries used

```{r libraries, echo=TRUE}
loadlibs <- function(){
library(ggplot2)
library(reshape)
library(gplots)
library(edgeR)
library(CHBUtils)
library(grid)
library(gridExtra)
library(Biobase)
library(genefilter)
library(RColorBrewer)
library(DESeq2)
library(SpeCond)
library(ggdendro)
library(vsn)
library(pheatmap)
}
suppressPackageStartupMessages(loadlibs())
```

### Get variables
- get base directory for analyses
- specify data and results directories
- specify column headers used in metadata file

```{r directories, echo=TRUE}
baseDir=getwd()
dataDir=paste(baseDir, "/combined-data-2", sep="")
resultsDir=paste(baseDir, "/results-combined-data", sep="")

heatcolors.1 <- rev(brewer.pal(6, "YlOrRd"))
heatcolors.2 <- colorRampPalette(brewer.pal(9, "GnBu"))(100)
cbPalette <- cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", 
                            "#D55E00", "#CC79A7", "#000000")
```

```{r functions, echo=FALSE}

getDendro <- function(x, y) {
  require(ggdendro)
  meta.x <- meta
  myDist <- dist(t(mat))
  myTree <-hclust(myDist)
  dhc <- as.dendrogram(myTree)
  ddata <- dendro_data(dhc, type="rectangle")
  ddata$labels <- merge(ddata$labels, meta.x, by.x="label", by.y="row.names")
  return(ddata)
  }

```


## Load data
Count data is loaded in along with the metadata for each sample. Here we will remove the outlier sample Npy2r3 which displayed an obvious batch effect. We will create an eset object for downstream analysis.

```{r load data}
# Load data
counts <- read.delim(file.path(dataDir, 'combined.counts'), sep="\t", row.names=1, header=T)
ann.counts <- read.delim(file.path(dataDir, 'annotated_combined.counts'), sep="\t", row.names=1, header=T)
meta <- read.csv(file.path(dataDir, 'project-summary.csv'), row.names=1, header=T)
row.names(meta) <- gsub("-", ".", row.names(meta))
meta <- meta[colnames(counts),]

# Remove outlier sample
outlier <- "Npy2r3"
meta <- meta[which(row.names(meta) != outlier),]
meta <- meta[,c('neurontype', 'replicate')]
meta$samplename = row.names(meta)


# Add batch information
batch <- read.delim('additional info/Batch_Info.csv', sep="\t", row.names=1, header=T)
batch <- batch[row.names(meta),]
meta <- cbind(meta, batch[,c('Lane', 'FCID')])

counts <- counts[,which(colnames(counts) %in% rownames(meta))]

# Create eset
eset <- new("ExpressionSet", exprs=as.matrix(counts))
pData(eset) <- meta
fData(eset) <- data.frame(GeneSymbol=ann.counts[,'geneSymbol'], row.names=row.names(ann.counts))

```

## Clustering of Data

### Sample-to-sample correlation matrix
To visualize clustering of data we will use transformed data by first running the count matrix through [DESeq2](https://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.pdf). For certain comparison analysis using unsupervised techniques, it is useful to transform data. These techniques (i.e. PCA and clustering) perform better when values have a similar dynamic range. Transformation renders data homoskedastic (variance of gene is stabilized across all expression levels), which you don't see with a simple log2 transform. Here, we use the rlog transform which takes into account gene-wise dispersion estimates and size factors (since the coverage varies quite a bit across samples). 

An intercorrelation heatmap is plotted below generated by taking a correlation of transformed values for all pairwise combinations of samples. The samples show high correlations with each other (values higher than 0.98) indicating no major outliers. **The clustering pattern is interesting, with three major groups of well clustered subtypes with the exception of SA1.2 and Npy2r2. There also seems to be no clustering associated with flowcell (FCID) or with Lane.** The three groups are as follows:   

* Group 1: Peptidergic, Non-peptidergic, C-LTMR
* Group 2: Proprioceptor, SA1
* Group 3: TrkB, TrkC, Npy2r


```{r cluster-ica, fig.align='center', echo=FALSE, warning=FALSE, message=FALSE}

# Create DESeq2 dataset
dds <- DESeqDataSetFromMatrix(countData = counts, colData = meta, design = ~ neurontype)
dds <- DESeq(dds)

# Matrix of pseudocounts for downstream visualization: two methods
rld <- rlog(dds, blind = TRUE)

# Heatmap of sample-to-sample correlation matrix
annotation <- data.frame(meta[,c('neurontype', 'Lane', 'FCID')], row.names=row.names(meta))
Lane <- c("black", "darkgrey", "white")
neurontype <- brewer.pal(8, "Paired")
names(neurontype) <- levels(meta$neurontype)
names(Lane) <- levels(meta$Lane)
ann_colors = list(Lane=Lane, neurontype=neurontype)

pheatmap(cor(assay(rld)), color = heatcolors.1, cluster_rows = T, 
         annotation=annotation, border_color=NA, annotation_colors=ann_colors,
         cluster_cols = T, show_colnames = T, clustering_distance_rows = "euclidean", 
         clustering_distance_cols = "euclidean", 
         fontsize = 10, fontsize_row = 10, height=20)
```


### PCA of samples
Using the **500 most variable genes**, we plotted a PCA plot of the first two principal components where combined ~60% of the variance is explained. The different neuronal subtypes appear to segregate well with the exception of Proprioceptor and RA-LTMR (Npy2r) which tend to cluster together


```{r pca, warning=FALSE, fig.align='center', echo=FALSE}

# Perform PCA (requires transformed data matrix)
ntop <- 500
rv <- rowVars(assay(rld))
select <- order(rv, decreasing=TRUE)[seq_len(ntop)]
pca <- prcomp(t(assay(rld)[select,]))

# Create data frame for input to ggplot
df <- cbind(meta, pca$x[,c('PC1', 'PC2')])

# Plot with sample names used as data points
ggplot(df, aes(PC1, PC2, color = neurontype)) + 
  theme_bw() +
  geom_point(size=4) +
  xlab('PC1 (50.45% variance explained)') +
  ylab('PC2 (11.82% variance explained)') +
  scale_x_continuous(expand = c(0.3,  0.3)) +
  theme(plot.title = element_text(size = rel(1.5)),
        axis.title = element_text(size = rel(1.5)),
        axis.text = element_text(size = rel(1.25)))

```

### Dendrogram based on euclidean distance of individual samples

```{r cluster-dendro, echo=FALSE, warning=FALSE, fig.height=10, fig.width=20, fig.align='center'}
mat <- assay(rld)[select,]
ddata <- getDendro(x=meta, y=mat)

# Full dataset
ggplot(segment(ddata)) +
    geom_segment(aes(x=x, y=y, xend=xend, yend=yend)) +
    theme_dendro() +
    geom_text(data=label(ddata), aes(x=x, y=y, label=label, color= label(ddata)[['neurontype']], 
                                     hjust=-0.1), size=6) +
    coord_flip() + scale_y_reverse(expand=c(0.2, 50)) +
    ggtitle('Dendrogram based on 500 most variable genes') +
    theme(axis.text.x=element_blank(),
          axis.text.y=element_blank(),
          axis.title.x=element_blank(),
          axis.title.y=element_blank(),
          legend.title=element_blank())
```

### Gene-set specific clustering
Here, we use different subsets of genes to generate dendrograms and illustrate the clustering of samples. What we see is that for the most part clustering does not change with gene set. That is, samples group as neuronal subtypes into clades but the positioning of that clade changes between gene groups.  The only exception is the GPCR gene set, where the SA1 samples don't all group into a single clade.

```{r cluster-genelist, echo=FALSE, warning=FALSE, fig.width=15, fig.height= 20, fig.align='center'}

# Get gene list 
genelist <- read.delim("additional info//gene_lists.txt", sep="\t", header=T)

# GPCRs
gpcr <- genelist$Gene[which(genelist$Family == "GPCRs")]
select <- row.names(ann.counts)[which(ann.counts$geneSymbol %in% gpcr)]
mat <- assay(rld)[select,]
ddata <- getDendro(x=meta, y=mat)

p1 <-  ggplot(segment(ddata)) +
    geom_segment(aes(x=x, y=y, xend=xend, yend=yend)) +
    theme_dendro() +
    geom_text(data=label(ddata), aes(x=x, y=y, label=label, color= label(ddata)[['neurontype']], 
                                     hjust=-0.1), size=6) +
    coord_flip() + scale_y_reverse(expand=c(0.2, 50)) +
    ggtitle('GPCR family genes') +
    theme(axis.text.x=element_blank(),
          axis.text.y=element_blank(),
          axis.title.x=element_blank(),
          axis.title.y=element_blank(),
          legend.title=element_blank())

# Neurotransmitter Receptors
nt <- genelist$Gene[which(genelist$Family == "NT receptors")]
select <- row.names(ann.counts)[which(ann.counts$geneSymbol %in% nt)]
mat <- assay(rld)[select,]
ddata <- getDendro(x=meta, y=mat)

p2 <-  ggplot(segment(ddata)) +
    geom_segment(aes(x=x, y=y, xend=xend, yend=yend)) +
    theme_dendro() +
    geom_text(data=label(ddata), aes(x=x, y=y, label=label, color= label(ddata)[['neurontype']], 
                                     hjust=-0.1), size=6) +
    coord_flip() + scale_y_reverse(expand=c(0.2, 50)) +
    ggtitle('Neurotransmitter receptor genes') +
    theme(axis.text.x=element_blank(),
          axis.text.y=element_blank(),
          axis.title.x=element_blank(),
          axis.title.y=element_blank(),
          legend.title=element_blank())

# VGIC
vgic <- genelist$Gene[which(genelist$Family == "VGIC")]
select <- row.names(ann.counts)[which(ann.counts$geneSymbol %in% vgic)]
mat <- assay(rld)[select,]
ddata <- getDendro(x=meta, y=mat)

p3 <-  ggplot(segment(ddata)) +
    geom_segment(aes(x=x, y=y, xend=xend, yend=yend)) +
    theme_dendro() +
    geom_text(data=label(ddata), aes(x=x, y=y, label=label, color= label(ddata)[['neurontype']], 
                                     hjust=-0.1), size=6) +
    coord_flip() + scale_y_reverse(expand=c(0.2, 50)) +
    ggtitle('VGIC genes') +
    theme(axis.text.x=element_blank(),
          axis.text.y=element_blank(),
          axis.title.x=element_blank(),
          axis.title.y=element_blank(),
          legend.title=element_blank())

# All Channels
channels <- genelist$Gene[which(genelist$Family == "All Channels")]
select <- row.names(ann.counts)[which(ann.counts$geneSymbol %in% channels)]
mat <- assay(rld)[select,]
ddata <- getDendro(x=meta, y=mat)

p4 <-  ggplot(segment(ddata)) +
    geom_segment(aes(x=x, y=y, xend=xend, yend=yend)) +
    theme_dendro() +
    geom_text(data=label(ddata), aes(x=x, y=y, label=label, color= label(ddata)[['neurontype']], 
                                     hjust=-0.1), size=6) +
    coord_flip() + scale_y_reverse(expand=c(0.2, 50)) +
    ggtitle('All Channel genes') +
    theme(axis.text.x=element_blank(),
          axis.text.y=element_blank(),
          axis.title.x=element_blank(),
          axis.title.y=element_blank(),
          legend.title=element_blank())

grid.arrange(p1,p2,p3,p4, ncol=2)
```


## DESeq2: Neuron versus All comparison
We set up the contrasts to assses gene expression changes for each neurontype against all other neurontypes. Significant genes are identfied as **log2 FC > 2 and corrected p-value < 0.05**. 

```{r deseq2-analysis, fig.align='center'}

# Create list object for the results 
resultsAll <- vector("list", 8)

# Extract data of specified contrasts; each neuron against everything else
# Start at 2 because the first element is the intercept
for (n in 2:length(resultsNames(dds))){
  contrast <- rep(0, length(resultsNames(dds)))
  contrast[n] <- 1
  res <- results(dds, contrast=contrast) # get results for neuron against all samples
  resultsAll[n-1] <- list(res)
}
names(resultsAll) <- resultsNames(dds)[-1]

# Get significant genes from each comparison
p.cutoff <- 0.05
fc <- 2
sigmat <- sapply(resultsAll, function(x){
            gene.FDR <- as.logical(abs(x$log2FoldChange) > fc & x$padj < p.cutoff)
            return(gene.FDR)}, USE.NAMES=F)
row.names(sigmat) <- row.names(exprs(eset))
colnames(sigmat) <- gsub("neurontype", "", colnames(sigmat))

```

### Significant genes
The number of genes in each list vary, with a fairly large number of genes (> 1k) being represented with SA1 and Nonpeptidergic subtypes. We also have listed the number of those significant genes which only appear for a single neuronal subtype.

```{r deseq2-sigtable, results='asis', echo=FALSE, fig.align='center'}
sigout <- cbind(colnames(sigmat), unname(apply(sigmat, 2, function(x){length(which(x))})))

# Get unique genes for each subtype
unique_genes <- numeric()

for (m in 1:nrow(sigout)){  
  geneset.1 <- row.names(sigmat)[which(sigmat[,m])]  
  
  for(n in 1:nrow(sigout)){
     if(n == m) next
     geneset.2 <- row.names(sigmat)[which(sigmat[,n])]
     geneset.1 <- geneset.1[which(geneset.1 %in% geneset.2 == FALSE)]
  }
  unique_genes <- c(unique_genes, length(geneset.1))
}

sigout <- cbind(sigout, unique_genes)
colnames(sigout) <-c("Neurontype", "No. significant genes", "No. of unique genes")
kable(sigout, format="markdown", row.names=F)
```


## Comparison of genes across subtypes
The table below summarizes the overlap between gene sets identified for each neuronal subtype. There are some pairs of subtypes that show a particularly large overlap, but doesn't necessarily coincide with the clustering we have observed earlier in this report.

```{r comparison, results='asis', echo=FALSE}
overlap_matrix <- matrix(,ncol=8, nrow=8)

for (m in 1:nrow(sigout)){  
  geneset.1 <- row.names(sigmat)[which(sigmat[,m])]  
  overlaps <- numeric()
  for(n in 1:nrow(sigout)){
     geneset.2 <- row.names(sigmat)[which(sigmat[,n])]
     overlaps <- c(overlaps, length(which(geneset.1 %in% geneset.2)))
  }
  overlap_matrix[,m] <- overlaps
}

colnames(overlap_matrix) <- colnames(sigmat)
rownames(overlap_matrix) <- paste(colnames(sigmat), "(", sigout[,2], ")", sep="")

kable(overlap_matrix, format="markdown")
```

## Files for download

* [Full count data matrix](./combined-data-2/annotated_combined.counts)
* [Full FPKM matrix](./combined-data-2/annotated_combined.fpkm)
* [A-LTMR Unique gene list](./combined-data-2/results/unique_gene_lists/A.LTMR_unique.txt)
* [AB (WDR) Unique gene list](./combined-data-2/results/unique_gene_lists/AB.WDR._unique.txt)
* [C-LTMR Unique gene list](./combined-data-2/results/unique_gene_lists/C.LTMR_unique.txt)  
* [Nonpeptidergic Unique gene list](./combined-data-2/results/unique_gene_lists/Nonpeptidergic.Nociceptor_unique.txt)  
* [Peptidergic Unique gene list](./combined-data-2/results/unique_gene_lists/Peptidergic.Nociceptor_unique.txt)  
* [RA-LTMR Unique gene list](./combined-data-2/results/unique_gene_lists/RA.LTMR_unique.txt)  
* [SA1 Unique gene list](./combined-data-2/results/unique_gene_lists/SA1_unique.txt) 
* [Proprioceptor Unique gene list](./combined-data-2/results/unique_gene_lists/Proprioceptor_unique.txt)
  
```{r write-to-file, echo=FALSE, eval=FALSE}

for (m in 1:nrow(sigout)){  
  geneset.1 <- row.names(sigmat)[which(sigmat[,m])]  
  
  for(n in 1:nrow(sigout)){
     if(n == m) next
     geneset.2 <- row.names(sigmat)[which(sigmat[,n])]
     geneset.1 <- geneset.1[which(geneset.1 %in% geneset.2 == FALSE)]
  }
  unique_genes <- geneset.1
  mset <- match(unique_genes, row.names(eset))
  unique_symbols <- as.character(fData(eset)$GeneSymbol[mset])
  write.table(cbind(unique_genes, unique_symbols), file=paste("combined-data-2/results/", sigout[m,1], "_unique.txt", sep=""),
              row.names=F, sep="\t", quote=F)
}


```



